data/transactions/logicsig.go

```diff
@@ -0,0 +1,93 @@
+// Copyright (C) 2019 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package transactions
+
+import (
+	"errors"
+
+	"github.com/algorand/go-algorand/crypto"
+	"github.com/algorand/go-algorand/protocol"
+)
+
+// LogicSig contains logic for validating a transaction.
+// LogicSig is signed by an account, allowing delegation of operations.
+// OR
+// LogicSig defines a contract account.
+type LogicSig struct {
+	_struct struct{} `codec:",omitempty,omitemptyarray"`
+
+	// Logic is an expression in a TBD language.
+	// Logic signed by Sig or Msig, OR hashed to be the Address of a contract account.
+	Logic []byte `codec:"lgc"`
+
+	Sig  crypto.Signature   `codec:"sig"`
+	Msig crypto.MultisigSig `codec:"msig"`
+
+	// Args are not signed, but checked by Logic
+	Args [][]byte `codec:"arg"`
+}
+
+// Blank returns true if there is no content in this LogicSig
+func (lsig *LogicSig) Blank() bool {
+	return len(lsig.Logic) == 0
+}
+
+// ToBeHashed implements our crypto.Hashable interface
+func (lsig *LogicSig) ToBeHashed() (protocol.HashID, []byte) {
```

Comment:
Should this be a hash of `Logic` rather than the entire `LogicSig`?  (That is, this method receiver should be on type `*Logic`.)

Reply:
I was trying to avoid making nested structures right away, like Txn/SignedTxn, but that might be better in the long run.
It'd be
`SignedLogic {
Logic { Logic: []byte },
Sig, Msig}`

Reply:
Doesn't need to be a nested structure, you can define `type LogicCode []byte` and then define a method receiver on `LogicCode`.

Reply:
To clarify why I think this is a good idea: our convention is that `Hashable` hashes the whole thing on which `Hashable` is called, but here, `Hashable` isn't returning the entire struct.

Reply:
Thanks.  Just curious why you didn't like the type alias for `[]byte`?  It would avoid the inefficiency you mentioned in the commit.